contracts/Eip712Forwrader.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier:MIT
+pragma solidity ^0.6.2;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import "./utils/GsnUtils.sol";
+
+// a Generic EIP712 forwarder.
+// actual struct has to START with known fields, but may contain other fields
+contract Eip712Forwarder {
+    using ECDSA for bytes32;
+
+    //all valid requests must start with this prefix.
+    // (note that the prefix is not a valid type by itself: it doesn't end with ")" (or ",", if more fields are added)
+    string constant typehashPrefix = "Forward(address sender, address target, bytes encodedFunction, uint256 gasLimit, address forwarder";
+
+    mapping(bytes32 => bool) public typeHashes;
+
+    struct Forward {
+        address sender;
+        address target;
+        bytes encodedFunction;
+        uint256 gasLimit;
+        uint256 nonce;
+        address forwarder;
+    }
+
+    // Nonces of senders, used to prevent replay attacks
+    mapping(address => uint256) private nonces;
+
+    function getNonce(address from) external view returns (uint256) {
+        return nonces[from];
+    }
+
+    function verify(Forward memory req,
+        bytes32 domainSeparator, bytes32 requestTypeHash, bytes memory suffixData, bytes memory sig) public view {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+    }
+
+    function verifyAndCall(Forward memory req,
+        bytes32 domainSeparator, bytes32 requestTypeHash, bytes memory suffixData, bytes memory sig) public {
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+        _updateNonce(req);
+
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory returnValue) = req.target.call{gas : req.gasLimit}(abi.encodePacked(req.encodedFunction, req.sender));
```

Comment:
I think it is forwarder's responsibility to make sure exactly 'gasLimit' gas is available to the recipient code. So should verify that `gasleft - gaslimit > 1/64` or whatever, don't you think this is a reasonable expectation from the forwarder with signed gas limit?

Reply:
yes, gas limit should be here. but I don't think we should put into it the EVM-specific calculation. passing a value as-is is good enough